code/modules/recycling/disposal.dm

```diff
@@ -675,258 +675,259 @@
-		// Leaving it intact and sitting in a wall is stupid.
-		if(T.density)
+	if(!T.is_plating() && istype(T,/turf/simulated/floor)) //intact floor, pop the tile
+		var/turf/simulated/floor/F = T
+		F.break_tile()
+		new /obj/item/stack/tile(H)	// add to holder so it will be thrown with other stuff
+
+	var/turf/target
+	if(direction)		// direction is specified
+		if(istype(T, /turf/space)) // if ended in space, then range is unlimited
+			target = get_edge_target_turf(T, direction)
+		else						// otherwise limit to 10 tiles
+			target = get_ranged_target_turf(T, direction, 10)
+
+		playsound(src, 'sound/machines/hiss.ogg', 50, 0, 0)
+		if(H)
 			for(var/atom/movable/AM in H)
-				AM.loc = T
-				AM.pipe_eject(0)
+				AM.forceMove(T)
+				AM.pipe_eject(direction)
+				spawn(1)
+					if(AM)
+						AM.throw_at(target, 100, 1)
+			H.vent_gas(T)
 			qdel(H)
-			return
 
+	else	// no specified direction, so throw in random direction
 
-		if(!T.is_plating() && istype(T,/turf/simulated/floor)) //intact floor, pop the tile
-			var/turf/simulated/floor/F = T
-			F.break_tile()
-			new /obj/item/stack/tile(H)	// add to holder so it will be thrown with other stuff
+		playsound(src, 'sound/machines/hiss.ogg', 50, 0, 0)
+		if(H)
+			for(var/atom/movable/AM in H)
+				target = get_offset_target_turf(T, rand(5)-rand(5), rand(5)-rand(5))
 
-		var/turf/target
-		if(direction)		// direction is specified
-			if(istype(T, /turf/space)) // if ended in space, then range is unlimited
-				target = get_edge_target_turf(T, direction)
-			else						// otherwise limit to 10 tiles
-				target = get_ranged_target_turf(T, direction, 10)
+				AM.forceMove(T)
+				AM.pipe_eject(0)
+				spawn(1)
+					if(AM)
+						AM.throw_at(target, 5, 1)
 
-			playsound(src, 'sound/machines/hiss.ogg', 50, 0, 0)
-			if(H)
-				for(var/atom/movable/AM in H)
-					AM.forceMove(T)
-					AM.pipe_eject(direction)
-					spawn(1)
-						if(AM)
-							AM.throw_at(target, 100, 1)
-				H.vent_gas(T)
-				qdel(H)
-
-		else	// no specified direction, so throw in random direction
-
-			playsound(src, 'sound/machines/hiss.ogg', 50, 0, 0)
-			if(H)
-				for(var/atom/movable/AM in H)
-					target = get_offset_target_turf(T, rand(5)-rand(5), rand(5)-rand(5))
+			H.vent_gas(T)	// all gas vent to turf
+			qdel(H)
 
-					AM.forceMove(T)
-					AM.pipe_eject(0)
-					spawn(1)
-						if(AM)
-							AM.throw_at(target, 5, 1)
+	return
 
-				H.vent_gas(T)	// all gas vent to turf
-				qdel(H)
+// call to break the pipe
+// will expel any holder inside at the time
+// then delete the pipe
+// remains : set to leave broken pipe pieces in place
+/obj/structure/disposalpipe/proc/broken(var/remains = 0)
+	if(remains)
+		for(var/D in GLOB.cardinal)
+			if(D & dpdir)
+				var/obj/structure/disposalpipe/broken/P = new(src.loc)
+				P.set_dir(D)
+
+	src.set_invisibility(101)	// make invisible (since we won't delete the pipe immediately)
+	var/obj/structure/disposalholder/H = locate() in src
+	if(H)
```

Comment:
There are a LOT of places here where a disposalholder can get deleted

this should check if (!QDELETED(H)) to ensure its not getting deleted multiple times